src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtils.java

```diff
@@ -51,8 +49,49 @@
      */
     static JavaRDD<GATKRead> transformReads(final SAMFileHeader header, final MarkDuplicatesScoringStrategy scoringStrategy, final OpticalDuplicateFinder finder, final JavaRDD<GATKRead> reads, final int parallelism) {
 
-        JavaPairRDD<String, Iterable<GATKRead>> keyedReads =
-                reads.mapToPair(read -> new Tuple2<>(ReadsKey.keyForRead(header, read), read)).groupByKey(parallelism);
+        JavaPairRDD<String, Iterable<GATKRead>> keyedReads;
+
+        if (SAMFileHeader.SortOrder.queryname.equals(header.getSortOrder())) {
+            // Avoid doing a shuffle since the reads are already name sorted.
+            // Find group boundaries by checking when the read key changes.
+            keyedReads = reads.mapPartitionsToPair(readi -> {
+                final PeekingIterator<GATKRead> iter = Iterators.peekingIterator(readi);
+                return new Iterable<Tuple2<String, Iterable<GATKRead>>>() {
+                    @Override
+                    public Iterator<Tuple2<String, Iterable<GATKRead>>> iterator() {
+                        return new AbstractIterator<Tuple2<String, Iterable<GATKRead>>>() {
+                            @Override
+                            protected Tuple2<String, Iterable<GATKRead>> computeNext() {
+                                String key = null;
+                                List<GATKRead> group = Lists.newArrayList();
+                                while (iter.hasNext()) {
+                                    if (key == null) {
+                                        GATKRead read = iter.next();
+                                        key = ReadsKey.keyForRead(header, read);
+                                        group.add(read);
+                                        continue;
+                                    }
+                                    GATKRead read = iter.peek();
+                                    String nextKey = ReadsKey.keyForRead(header, read);
+                                    if (nextKey.equals(key)) {
+                                        group.add(iter.next());
+                                    } else {
+                                        return new Tuple2<>(key, group);
+                                    }
+                                }
+                                if (key != null) {
+                                    return new Tuple2<>(key, group);
+                                }
+                                return endOfData();
+                            }
+                        };
+                    }
+                };
+            });
```

Comment:
There may be a more compact way of doing this grouping with Java 8 or Guava libraries. If anyone knows of a way I'm all ears.